bank: create and delete account functionality

BorislavAnchev opened the pull request:
The user is now able to add a new bank account on demand as well as delete from his current bank accounts.

BorislavAnchev commented:
@mgenov, new code ready for a review.

mgenov commented on banktask/src/redux/modules/account/reducers.spec.js:
```diff
@@ -95,4 +95,81 @@ describe('Accounts reducer', () => {
     });
   });
+
+  describe('case: TYPES.DELETE_ACCOUNTS_SUCCESS', () => {
```
doesn't look reasonable naming as you add a describe block for a single scenario.

mgenov commented on banktask/src/redux/modules/account/reducers.spec.js:
```diff
@@ -95,4 +95,81 @@ describe('Accounts reducer', () => {
+          alert: 'Account deleted successfully!'
+        }
+      }
+      const newState = accountsReducer(fakeInitialState, fakeAction);
```
It would be more readable if you pass the initial state and action directly in the reducer.

mgenov commented on banktask/src/redux/modules/account/reducers.spec.js:
```diff
@@ -95,4 +95,81 @@ describe('Accounts reducer', () => {
+          id: '_u70nyuzcq'
+        },
+        meta: {
+          alert: 'Account deleted successfully!'
```
Doesn't look reasonable to be added in the delete action as this is the alerts are related to UI.

mgenov commented on banktask/src/App.js:
```diff
@@ -18,6 +19,7 @@ const App = () => {
       <div className="App">
         <Header data-test='Header Component'/>
         <Route exact path='/' render={() => <HomePage />} data-testid='Home Page Route' />
+        <Route exact path='/create' render={(props) => <CreateAccount {...props}/>} data-testid='Create Account Page Route' />
```
/account/new 

Could be named NewAccountPage if we follow the HomePage convention.

mgenov commented on banktask/src/App.test.js:
```diff
@@ -27,4 +27,10 @@ describe('App component', () => {
     expect(getByTestId('Home Page Component')).toBeInTheDocument();
   });
+
+  it('Should render a Create account page component', () => {
```
There is a duplication in the Provider component that could be avoided using wrapper of the render call. 

Here is an example that could you help you remove this duplication:

```jsx
function render(
  ui,
  {
    initialState,
    store = createStore(dummyReducer, initialState),
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] })
  } = {}
) {
  return {
    ...rtlRender(
      <Provider store={store}>
        <Router history={history}>{ui}</Router>
      </Provider>
    ),
    // adding `store` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    store
  }
}
```

mgenov commented on banktask/src/mocks/backend.js:
```diff
@@ -135,4 +136,15 @@ mock.onPut('/accounts')
+});
+
+mock.onDelete('/accounts').reply(({params}) => {
+  return [200, { id: params.id }]
```
Doesn't look like it deletes something from the in-memory database.

BorislavAnchev commented:
@mgenov, new code ready for a review:

- Removed the `describe()` blocks for the reducer scenarios
- The initial state and the actions are now directly passed to the reducer
- Removed the 'meta' field from the fake actions as well as from the real action creators
- The 'CreateAccount' page is now renamed to 'NewAccountPage' and the path to the page is changed to '/account/new'
- Implemented the custom render function
- The back-end mock now removes the database entry with the specified ID on DELETE request

mgenov commented on banktask/src/screens/NewAccountPage/NewAccountPage.spec.js:
```diff
@@ -0,0 +1,79 @@
+
+  afterEach(cleanup);
+
+  it('Should render an account IBAN input field component', () => {
```
Most of the statements are with "Should render" that could be wrapped into a describe block for readability. 

```jsx
describe('(CreateAccount) component', () => {
  describe('is rendering', () => {
      it('iban button', () => {

      })

      it('iban title', () => {
      })
  })
})
```

mgenov commented on banktask/src/screens/NewAccountPage/NewAccountPage.spec.js:
```diff
@@ -0,0 +1,79 @@
+    expect(getByTestId('Back Button Primitive')).toBeInTheDocument();
+  });
+
+  describe('Account create warning', () => {
```
Creation of warning sounds misleading as warnings are displayed to the users.

BorislavAnchev commented:
Changes ready for review:

- Tests of component rendering are now in a separate describe block
- Changed the description of the account creation warning

BorislavAnchev commented:
@mgenov, bumping this just to be sure.

mgenov commented on banktask/src/screens/NewAccountPage/NewAccountPage.js:
```diff
@@ -0,0 +1,70 @@
+  const [addAccountWarning, setAddAccountWarning] = useState('');
+  const accounts = useSelector((state) => state.accounts);
+
+  const currencyOptions = ['BGN', 'USD', 'EUR'];
```
Could be moved it separate component for re-usability.

mgenov commented:
You can add screenshots in the body of the PR for clearance how the things are rendered in the UI.

BorislavAnchev commented:
@mgenov, changes ready for a review:

- Moved the currency options array into the utility module
- Created demo GIF for how the UI is rendered

![Bank 2019-11-27 09-55](https://user-images.githubusercontent.com/51535566/69704781-d5068a00-10fc-11ea-9cb8-82c3d9105e43.gif)

BorislavAnchev commented:
@mgenov, bumping again.

BorislavAnchev commented:
@mgenov, changes ready for a review.

BorislavAnchev commented:
@mgenov, am I expected to do any more corrections on this PR that I am unaware of? Because I may be waiting for a review while you are expecting me to do something else.

mgenov commented on banktask/src/utils/index.js:
```diff
@@ -1,6 +1,38 @@
+  }
+}
+
+export const currencyOptions = ['BGN', 'USD', 'EUR'];
```
it's not a good idea to keep currencyOptions in src/utils/index.js
